keyboards/darkproject/kd83a_bfg_edition/kd83a_bfg_edition.c

```diff
@@ -0,0 +1,257 @@
+    {  55,  22}, //  34, k42, R
+    {  66,  22}, //  35, k52, T
+    {  77,  22}, //  36, k62, Y
+    {  88,  22}, //  37, k72, U
+    {  99,  22}, //  38, k82, I
+    { 110,  22}, //  39, k92, O
+    { 121,  22}, //  40, ka2, P
+    { 132,  22}, //  41, kb2, [
+    { 143,  22}, //  42, kc2, ]
+    { 154,  22}, //  43, kd3, "\\"
+    { 165,  22}, //  44, k06, Del
+
+    {  11,  33}, //  45, k03, Caps Lock
+    {  22,  33}, //  46, k13, A
+    {  33,  33}, //  47, k23, S
+    {  44,  33}, //  48, k33, D
+    {  55,  33}, //  49, k43, F
+    {  66,  33}, //  50, k53, G
+    {  77,  33}, //  51, k63, H
+    {  88,  33}, //  52, k73, J
+    {  99,  33}, //  53, k83, K
+    { 110,  33}, //  54, k93, L
+    { 121,  33}, //  55, ka3, ;
+    { 132,  33}, //  56, kb3, '
+    { 154,  33}, //  57, kc3, Enter
+    { 165,  33}, //  58, k26, PgUp
+   
+    {  11,  44}, //  59, k04, Shift_L
+    {  33,  44}, //  60, k24, Z
+    {  44,  44}, //  61, k34, X
+    {  55,  44}, //  62, k44, C
+    {  66,  44}, //  63, k54, V
+    {  77,  44}, //  64, k64, B
+    {  88,  44}, //  65, k74, N
+    {  99,  44}, //  66, k84, M
+    { 110,  44}, //  67, k94, ,
+    { 121,  44}, //  68, ka4, .
+    { 132,  44}, //  69, kb4, /
+    { 143,  44}, //  70, kd4, Shift_R
+    { 154,  44}, //  70, k17, Up
+    { 165,  44}, //  72, k36, PgDn
+
+    {  11,  55}, //  73, k05, Ctrl_L
+    {  22,  55}, //  74, k15, Win_L
+    {  33,  55}, //  75, k25, Alt_L
+    {  77,  55}, //  76, k65, Space
+    { 110,  55}, //  77, k95, Alt_R
+    { 121,  55}, //  78, ka5, FN
+    { 132,  55}, //  79, kc5, Ctrl_R
+    { 143,  55}, //  80, k07, Left
+    { 154,  55}, //  81, k27, Down
+    { 165,  55}, //  82, k37, Right  
+     
+    {   0,  18}, // 83, Z1
+    {   0,  24}, // 84, Z2
+}, {
+    4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 
+    4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 
+    4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+    1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 
+    4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 
+    4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 
+    2, 2 
+} };
+#endif
+
+#ifdef EEPROM_ENABLE
+
+#include "spi_master.h"
+
+void spi_init(void) {
+    static bool is_initialised = false;
+    if (!is_initialised) {
+        is_initialised = true;
+
+        // Try releasing special pins for a short time
+        setPinInput(SPI_SCK_PIN);
+        setPinInput(SPI_MOSI_PIN);
+        setPinInput(SPI_MISO_PIN);
+
+        chThdSleepMilliseconds(10);
+
+        palSetPadMode(PAL_PORT(SPI_SCK_PIN), PAL_PAD(SPI_SCK_PIN), PAL_MODE_ALTERNATE(SPI_SCK_PAL_MODE) | PAL_OUTPUT_TYPE_PUSHPULL | PAL_OUTPUT_SPEED_HIGHEST | PAL_WB32_CURRENT_LEVEL3);
+        palSetPadMode(PAL_PORT(SPI_MOSI_PIN), PAL_PAD(SPI_MOSI_PIN), PAL_MODE_ALTERNATE(SPI_MOSI_PAL_MODE) | PAL_OUTPUT_TYPE_PUSHPULL | PAL_OUTPUT_SPEED_HIGHEST);
+        palSetPadMode(PAL_PORT(SPI_MISO_PIN), PAL_PAD(SPI_MISO_PIN), PAL_MODE_ALTERNATE(SPI_MISO_PAL_MODE) | PAL_OUTPUT_TYPE_PUSHPULL | PAL_OUTPUT_SPEED_HIGHEST);
+    }
+}
+
+#endif
```

Comment:
This allows for a default behavior to be defined, without the need for the user to specify the behavior.  This is very useful for the configurator which doesn't have a way to configure this behavior, currently.

```suggestion
#endif

#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        if (clockwise) {
            tap_code_delay(KC_VOLU, 10);
        } else {
            tap_code_delay(KC_VOLD, 10);
        }
    }
    return true;
}
#endif
```